src/edu/washington/escience/myria/operator/Apply.java

```diff
@@ -47,47 +57,152 @@
+          int[] iteratorIndexes = new int[emitEvaluators.size()];
+
+          for (int rowIdx = 0; rowIdx < inputTuples.numTuples(); ++rowIdx) {
+            // First, get all result counts for this row.
+            boolean emptyProduct = false;
+            for (int i = 0; i < resultCountColumns.size(); ++i) {
+              int resultCount = resultCountColumns.get(i).getInt(rowIdx);
+              resultCounts[i] = resultCount;
+              lastCumResultCounts[i] = cumResultCounts[i];
+              cumResultCounts[i] += resultCounts[i];
+              if (resultCount == 0) {
+                // If at least one evaluator returned zero results, the Cartesian product is empty.
+                emptyProduct = true;
+              }
+            }
+
+            if (!emptyProduct) {
+              // Initialize each iterator to its starting index.
+              Arrays.fill(iteratorIndexes, 0);
+              // Iterate over each element of the Cartesian product and append to output.
+              do {
+                for (int iteratorIdx = 0; iteratorIdx < iteratorIndexes.length; ++iteratorIdx) {
+                  int resultRowIdx =
+                      lastCumResultCounts[iteratorIdx] + iteratorIndexes[iteratorIdx];
+                  outputBuffer.appendFromColumn(
+                      iteratorIdx, resultColumns.get(iteratorIdx), resultRowIdx);
+                }
+              } while (!computeNextCombination(resultCounts, iteratorIndexes));
+            }
+          }
+        }
+      } else {
+        // We don't want to keep polling in a loop since this method is non-blocking.
+        break;
+      }
     }
+    // If we produced a full batch, return it, otherwise finish the current batch and return it.
+    return outputBuffer.popAny();
+  }
 
-    List<Column<?>> output = Lists.newLinkedList();
-    for (GenericEvaluator evaluator : emitEvaluators) {
-      output.add(evaluator.evaluateColumn(tb));
+  /**
+   * This method mutates {@link iteratorIndexes} on each call to yield the next element of the Cartesian product of
+   * {@link upperBounds} in lexicographic order. If all elements have been exhausted, it returns true, otherwise it
+   * returns false.
+   *
+   * @param upperBounds an immutable array of elements representing the sets we are forming the Cartesian product of,
+   *          where each set is of the form [0, i), where i is an element of {@link upperBounds}
+   * @param iteratorIndexes a mutable array of elements representing the current element of the Cartesian product
+   * @return if we have exhausted all elements of the Cartesian product
+   */
+  private boolean computeNextCombination(final int[] upperBounds, final int[] iteratorIndexes) {
+    boolean endOfIteration = false;
+    int lastIteratorPos = iteratorIndexes.length - 1;
+    // Count backward from the innermost iterator to the outermost.
+    for (int iteratorPos = lastIteratorPos; iteratorPos >= 0; --iteratorPos) {
```

Comment:
Yes there is, if we're generating the product in lexicographic order, which I want to do.